File: web/src/app/products/product.ts
// A product as the API describes it, with what it is to the viewer.
export interface Product {
  id: number;
  title: string;
  brand: string;
  maker: string;
  description: string;
  price: number;
  category: string;
  stock: number;
  photoUrl: string;
  seller: string;
  createdAt: string;
  updatedAt: string;
  likes: number;
  liked: boolean;
  mine: boolean;
}

// What a seller sends. The photo is the name the upload route returned.
export interface ProductDraft {
  title: string;
  brand: string;
  maker: string;
  description: string;
  price: number;
  category: string;
  stock: number;
  photo: string;
}

// A category of the shop and how many products it holds.
export interface Category {
  name: string;
  count: number;
}

// How the catalogue can be narrowed: a search text, a category, both.
export interface ProductFilter {
  q?: string;
  category?: string;
}

export interface UploadedPhoto {
  photo: string;
  url: string;
}

File: api/src/HomeMarket.Api/Models/Product.cs
namespace HomeMarket.Api.Models
{
    // Something for sale, listed by one account, liked by any number of
    // others, in one of the shop's categories, with the number left in
    // stock. The photo is either a public URL or the name of an uploaded
    // file served by the API.
    public class Product
    {
        public int Id { get; set; }
        public string Title { get; set; } = string.Empty;
        public string Brand { get; set; } = string.Empty;
        public string Maker { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public decimal Price { get; set; }
        public string Category { get; set; } = string.Empty;
        public int Stock { get; set; }
        public string Photo { get; set; } = string.Empty;
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
        public int SellerId { get; set; }
        public Account? Seller { get; set; }
        public List<Account> LikedBy { get; set; } = new();
    }
}
